aws_lambda_powertools/shared/headers_serializer.py

```diff
@@ -0,0 +1,111 @@
+
+        # Format 2.0 doesn't have multiValueHeaders or multiValueQueryStringParameters fields.
+        # Duplicate headers are combined with commas and included in the headers field.
+        combined_headers: Dict[str, str] = {}
+        for key, values in headers.items():
+            if isinstance(values, str):
+                combined_headers[key] = values
+            else:
+                combined_headers[key] = ", ".join(values)
+
+        return {"headers": combined_headers, "cookies": cookies}
+
+
+class MultiValueHeadersSerializer(BaseHeadersSerializer):
+    def serialize(self, headers: Dict[str, Union[str, List[str]]], cookies: List[str]) -> Dict[str, Any]:
+        """
+        When using REST APIs, headers can be encoded using the `multiValueHeaders` key on the response.
+        This is also the case when using an ALB integration with the `multiValueHeaders` option enabled.
+        The solution covers headers with just one key or multiple keys.
+
+        https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format
+        https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#multi-value-headers-response
+        """
+        payload: Dict[str, List[str]] = defaultdict(list)
+
+        for key, values in headers.items():
+            if isinstance(values, str):
+                payload[key].append(values)
+            else:
+                for value in values:
+                    payload[key].append(value)
+
+        if cookies:
+            payload.setdefault("Set-Cookie", [])
+            for cookie in cookies:
+                payload["Set-Cookie"].append(cookie)
+
+        return {"multiValueHeaders": payload}
+
+
+class SingleValueHeadersSerializer(BaseHeadersSerializer):
+    def serialize(self, headers: Dict[str, Union[str, List[str]]], cookies: List[str]) -> Dict[str, Any]:
+        """
+        The ALB integration has `multiValueHeaders` disabled by default.
+        If we try to set multiple headers with the same key, or more than one cookie, print a warning.
+
+        https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#respond-to-load-balancer
+        """
+        payload: Dict[str, Dict[str, str]] = {}
+        payload.setdefault("headers", {})
+
+        if cookies:
+            if len(cookies) > 1:
+                warnings.warn(
+                    "Can't encode more than one cookie in the response. Sending the last cookie only. "
+                    "Did you enable multiValueHeaders on the ALB Target Group?"
+                )
+
+            # We can only send one cookie, send the last one
+            payload["headers"]["Set-Cookie"] = cookies[-1]
+
+        for key, values in headers.items():
+            if isinstance(values, str):
+                payload["headers"][key] = values
+            else:
+                if len(values) > 1:
```

Comment:
@heitorlessa I'm seeing `TypeError: object of type 'NoneType' has no len()` when I have the following header set:

```
'Access-Control-Allow-Origin': None
```

Reply:
@jasmarc could you create a bug report issue on this please?

We can make a patch release as soon as we reproduce it.

Thanks a lot!!!

Reply:
> with my laptop now

While we await for a bug report and a fix by EOD for headers with `None` value, you can safely use an empty string instead of `None`.

Because we provide [native support for CORS in Event Handler](https://awslabs.github.io/aws-lambda-powertools-python/2.4.0/core/event_handler/api_gateway/#cors), you can do as follows:

> Test URL: https://xwgyz6mo2m.execute-api.eu-west-1.amazonaws.com/Prod/hello/

```python
from aws_lambda_powertools.event_handler import APIGatewayRestResolver, CORSConfig

cors = CORSConfig(allow_origin="")  # Create CORS configuration
app = APIGatewayRestResolver(cors=cors)  # Ensure CORS is properly configured for all routes


@app.get("/hello")
def hello():
    return {"hello": "world"}


def lambda_handler(event, context):
    return app.resolve(event, context)
```

**If your intent is to set an explicit `null` in the Allow-Origin**...

[W3C doesn't recommend having an Allow Origin explicitly set to `null`](https://w3c.github.io/webappsec-cors-for-developers/#avoid-returning-access-control-allow-origin-null) due to potential data leak, since `data:` and `file:` schemes can access`null` origins - that's not the case with an empty Allow Origin (`''`). Unless I'm misunderstanding both your intent (bug report helps!) or W3C docs.

---

With an explicit `null`

```python
import json

from aws_lambda_powertools.event_handler import (
    APIGatewayRestResolver,
    CORSConfig,
    Response,
)

# Not recommended due to data: and file: schemes also using `null` origin
# https://w3c.github.io/webappsec-cors-for-developers/#avoid-returning-access-control-allow-origin-null
cors = CORSConfig(allow_origin="null")
app = APIGatewayRestResolver(cors=cors)


# https://xwgyz6mo2m.execute-api.eu-west-1.amazonaws.com/Prod/hello/
@app.get("/hello")
def hello():
    return {"hello": "world"}


# chore: verify API Gateway REST Lambda integration behaviour with `None`
# API Gateway REST serializes`None` as an empty string (just like `null` in JS)
# need proper E2E to validate behaviour for API Gateway HTTP (v2), ALB, and Function URL resolvers

# https://xwgyz6mo2m.execute-api.eu-west-1.amazonaws.com/Prod/none/
@app.get("/none")
def hello():
    headers = {"X-Empty": None, "X-Null": "null"}
    response = {"hello": "world"}
    return Response(body=json.dumps(response), headers=headers, status_code=200)


def lambda_handler(event, context):
    return app.resolve(event, context)
```

Reply:
@leandrodamascena when you're back online, feel free to create a bug report if we don't hear from the customer within an hour. Leaving the tests and my endpoint above to save you from setting it all up besides E2E.

---

> **tests/functional/test_headers_serializer.py**

These tests reproduce the issue. What needs to be done is an E2E for each resolver to confirm whether the final value becomes `''` or fails (ALB is always the odd one) - this needs to be documented too later.

```python
def test_http_api_headers_serializer_with_null_values():
    serializer = HttpApiHeadersSerializer()
    payload = serializer.serialize(headers={"Foo": None}, cookies=[])
    assert payload == {"headers": {}, "cookies": []}


def test_multi_value_headers_serializer_with_null_values():
    serializer = MultiValueHeadersSerializer()
    payload = serializer.serialize(headers={"Foo": None}, cookies=[])
    assert payload == {"headers": {}, "cookies": []}


def test_single_value_headers_serializer_with_null_values():
    serializer = SingleValueHeadersSerializer()
    payload = serializer.serialize(headers={"Foo": None}, cookies=[])
    assert payload["headers"] == {}
```

Reply:
Hey @heitorlessa sorry for the delay. Yes, I worked around it and I did also notice that it was not recommended, so I almost hesitated to bring it up.

Reply:
I have filed https://github.com/awslabs/aws-lambda-powertools-python/issues/1791